### ISSUE TITLE
오타 제보 - PART 04. 컴퓨터 비전

### ISSUE THREAD

**Issue by hschoi0705** [Justin-A/DeepLearning101#6]
### 1. 코드
  - p.136, 주석 6번행 들여쓰기 오류
  - p.159, 코드 56번, out = F.avg_pool2d(out, 8s) -> out = F.avg_pool2d(out, 8)
  - p.176, 코드 7번, transforms.C[Enter]Crop(224) -> transforms.CenterCrop(224)
  

### 2. 내용
  - p.121, 마지막문단, 예상치못한 URL이 포함되어 있음(확인결과 오픈마켓 한돈 암돼지1등급이... ^^ㅋ)
  - p.124, 아래서 2번째 문단, "사각형 안의 픽셀 값을 뽑으면 'max pooling'\~" -> "사각형 안의 최대 픽셀 값을 뽑으면 'max pooling'\~"
  - p.134, code의 optimizer와 주석 내용 상이함
  - p.147, 위에서 4째줄, [그림 4-12] -> [그림 4-14]
  - p.147, 2번째 문단, [그림 4-13] -> [그림 4-15]
  - p.148, 그림 참고 2개, [그림 4-14] -> [그림 4-16]
  - p.152, 그림 참고, [그림 4-15] -> [그림 4-17]
  - p.153, 그림 참고, [그림 4-18] -> [그림 4-20]
  - p.155, 첫째줄 내용 확인 필요, [그림 4-20]을 확대한 부분이 Inception 모듈입니다. -> [그림 4-22]의 확대한 부분이 Inception 모듈입니다.
  - p.156, ResNet 설명 문단 그림 참고, [그림 4-22] -> [그림 4-24]
  - p.177, 코드 7번에 대한 설명이 CenterCrop에 대한 설명이 아니라 RandomResizedCrop에 대한 설명으로 보입니다.
  - p.177, 코드 8번에 대한 설명이 코드와 상이합니다.
  - p.184, pretrained model 학습에 대한 코드 설명은 있으나 해당되는 코드는 없습니다.(전체코드에는 있음)
  - p.189, 11번 코드 부분 optimizer 이하 부분이 2번 반복되고 있음
  
### 3. 질문
  - 예제4-6, ResNet18 pretrained model 학습시 전체 네트워크를 학습하는 것인지요? Transfer Learning 내용 소개에서는 weight freezing에 대한 설명과 함께 fully connected 부분만 학습하는 것으로 소개되는데 코드에서는 별다른 언급이 없는것 같아서 문의드립니다.

**Comment by Justin-A** [comment]
안녕하세요, Justin입니다.
문의해주신 내용에 대해 답변드립니다.
책 내용에 대해서 다양한 부분을 지적해주셔서 감사합니다.

1. 코드
- 본 책에서 다루는 코드들에 대해서, 모든 들여쓰기는 4칸씩 진행됩니다. 지적해주신 코드에는 2칸씩 들여쓰기가 적용된 것으로 보일 수 있을 것 같습니다. 4칸 들여쓰기가 적용되어야 합니다.
- 지적해 주신 내용처럼 해당 내용은 다음과 같이 수정되어야 합니다.
```python
out = F.avg_pool2d(out, 8)
```
- 지적해 주신 내용처럼 해당 내용은 다음과 같이 수정되어야 합니다.
```python
transforms.CenterCrop(224),
```

2. 내용
- 마지막 문단에 들어있는 지마켓 링크는 제거되어야 합니다.
- '사각형 안의 픽셀 중 최대값을 뽑으면' 으로 수정되어야 합니다.
- 지적해 주신 내용은 다음과 같이 변경되어야 합니다. 
BackPropagation을 통해 파라미터를 업데이터할때, 이용하는 옵티마이저를 정의해줍니다. 본 예제에서는 Adam 알고리즘을 이용하며, 파라미터를 업데이트할 때 반영될 학습률을 0.001로 설정합니다. 보통 Adam을 기본옵션으로 설정하며 학습률은 분석가 취향에 따라서 다르게 설정하기도 합니다.
- [그림4-12]에서 [그림4-14]으로 수정되어야 합니다.
- [그림4-13]에서 [그림4-15]으로 수정되어야 합니다.
- [그림4-14]에서 [그림4-16]으로 수정되어야 합니다.
- [그림4-15]에서 [그림4-17]으로 수정되어야 합니다.
- [그림4-18]에서 [그림4-20]으로 수정되어야 합니다.
- [그림4-20]에서 [그림4-22]으로 수정되어야 합니다.
- [그림4-22]에서 [그림4-24]으로 수정되어야 합니다.
- 코드는 위에서 언급한 것 처럼, CenterCrop으로 수정되어야하며, 코드 설명 내용은 다음과 같습니다.
'해당 이미지에 대해 중앙을 기준으로 224 * 224 크기로 이미지를 잘라내어 사이즈를 변경하는 것을 의미합니다.'
- 코드 설명 내용은 다음과 같습니다.
'해당 이미지를 256 * 256 크기로 변경하는 것을 의미합니다.'
- 해당 내용이 누락되어있습니다. 추가되어야 하는 내용은 다음과 같습니다.
```python
''' 11. IMAGENET 데이터로 미리 학습이 된 ResNet18 모델을 불러온 후 개미, 벌 이미지 데이터에 맞게 Fine Tuning 해보기 '''
model = models.resnet18(pretrained = True)    # (1)
num_ftrs = model.fc.in_features               # (2)
model.fc = nn.Linear(num_ftrs, 2)             # (3)
model = model.cuda()                          # (4)

optimizer = torch.optim.Adam(model.parameters(), lr = 0.0001)          # (5)
EPOCHS = 10                                                            # (6)
for epoch in range(1, EPOCHS + 1):                                     # (7)
    train(model, dataloaders["train"], optimizer, log_interval = 5)    # (8)
    test_loss, test_accuracy = evaluate(model, dataloaders["val"])     # (9)
    print("\n[EPOCH: {}], \tTest Loss: {:.4f}, \tTest Accuracy: {:.2f} % \n".format(
        epoch, valid_loss, valid_accuracy))
```
- 지적해주신대로 2번 반복되어있습니다. 1번만 작성되도록 수정이 필요합니다.

3. 질문
- 전체 네트워크를 학습하는 것이 맞습니다. 이론 내용에서는 Fully Connected 부분만 학습하도록 설명이 되어있지만, 분석가 입장에서 Conv 층도 학습시킬 수 있으며, 보통 학습시킬 때, 학습률(Learning Rate)를 매우 낮게 설정해서 학습시키기도 합니다. 본 예제에서는 이미 학습된 네트워크에 대해서 전체 구조를 추가로 학습시키는 Fine Tuning 과정으로 진행하였습니다.

많은 부분을 지적해주셔서 감사드립니다. 지난번 답변처럼, 2판 인쇄시 수정되어 문제가 없도록 하겠습니다.
